src/chainrules.jl

```diff
@@ -0,0 +1,98 @@
+        return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), ChainRulesCore.Zero(), ChainRulesCore.Composite{typeof(s)}(value=Δvproj))
+    end
+    return (vproj, pullback)
+end
+
+function ChainRulesCore.rrule(::typeof(projection_on_set), d::DefaultDistance, v::T, s::MOI.LessThan) where {T}
+    vproj = projection_on_set(d, v, s)
+    function pullback(Δvproj)
+        if vproj == v # if exactly equal, then constraint inactive
+            return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), Δvproj, ChainRulesCore.Composite{typeof(s)}(upper=zero(Δvproj)))
+        end
+        return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), zero(Δvproj), ChainRulesCore.Composite{typeof(s)}(upper=Δvproj))
+    end
+    return (vproj, pullback)
+end
+
+function ChainRulesCore.rrule(::typeof(projection_on_set), d::DefaultDistance, v::T, s::MOI.GreaterThan) where {T}
+    vproj = projection_on_set(d, v, s)
+    function pullback(Δvproj)
+        if vproj == v # if exactly equal, then constraint inactive
+            return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), Δvproj, ChainRulesCore.Composite{typeof(s)}(lower=zero(Δvproj)))
+        end
+        return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), zero(Δvproj), ChainRulesCore.Composite{typeof(s)}(lower=Δvproj))
+    end
+    return (vproj, pullback)
+end
+
+function ChainRulesCore.rrule(::typeof(projection_on_set), d::DefaultDistance, v::AbstractVector{T}, s::MOI.Reals) where {T}
+    vproj = projection_on_set(d, v, s)
+    function pullback(Δvproj)
+        return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), Δvproj, ChainRulesCore.DoesNotExist())
+    end
+    return (vproj, pullback)
+end
+
+function ChainRulesCore.rrule(::typeof(projection_on_set), d::DefaultDistance, v::AbstractVector{T}, s::MOI.Zeros) where {T}
+    vproj = projection_on_set(d, v, s)
+    function pullback(Δvproj)
+        return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), FillArrays.Zeros(length(v)), ChainRulesCore.DoesNotExist())
+    end
+    return (vproj, pullback)
+end
+
+function ChainRulesCore.rrule(::typeof(projection_on_set), d::DefaultDistance, v::AbstractVector{T}, s::S) where {T, S <: Union{MOI.Nonnegatives, MOI.Nonpositives}}
+    vproj = projection_on_set(d, v, s)
+    function pullback(Δvproj)
+        v̄ = zeros(eltype(Δvproj), length(Δvproj))
+        for i in eachindex(Δvproj)
+            if vproj[i] == v[i]
+                v̄[i] = Δvproj[i]
+            end
+        end
+        return (ChainRulesCore.NO_FIELDS, ChainRulesCore.DoesNotExist(), v̄, ChainRulesCore.DoesNotExist())
+    end
+    return (vproj, pullback)
+end
+
+function ChainRulesCore.rrule(::typeof(projection_on_set), d::Union{DefaultDistance, NormedEpigraphDistance}, v::AbstractVector{T}, s::MOI.SecondOrderCone) where {T}
+    vproj = projection_on_set(d, v, s)
+    t = v[1]
+    x = v[2:end]
+    norm_x = LinearAlgebra.norm2(x)
+    function pullback(Δv)
```

Comment:
better stacktraces if you name it for the primal function
```suggestion
    function projection_on_set_pullback(Δv)
```